test/test_utils.py

```diff
@@ -711,6 +715,72 @@ def test_unmarshal_tools_pcr_values(self):
         self.assertEqual(digs[1], b"\x00" * 20)
         self.assertEqual(digs[2], b"\xFF" * 32)
 
+    def test_credential_to_tools(self):
+        id_object_len = 40
+        encrypted_secret_len = 68
+
+        id_object_bytes = os.urandom(id_object_len)
+        encrypted_secret_bytes = os.urandom(encrypted_secret_len)
+
+        id_object = TPM2B_ID_OBJECT(credential=id_object_bytes)
+        encrypted_secret = TPM2B_ENCRYPTED_SECRET(secret=encrypted_secret_bytes)
+
+        credential_blob = credential_to_tools(id_object, encrypted_secret)
+
+        self.assertEqual(credential_blob[0:4], int(0xBADCC0DE).to_bytes(4, "big"))
+        self.assertEqual(credential_blob[4:8], int(1).to_bytes(4, "big"))
+        self.assertEqual(credential_blob[8:10], int(id_object_len).to_bytes(2, "big"))
+        self.assertEqual(credential_blob[10 : 10 + id_object_len], id_object_bytes)
+
+        self.assertEqual(
+            int.from_bytes(
+                credential_blob[10 + id_object_len : 12 + id_object_len],
+                byteorder="big",
+            ),
+            encrypted_secret_len,
+        )
+        self.assertEqual(
+            credential_blob[
+                12 + id_object_len : 12 + id_object_len + encrypted_secret_len
+            ],
+            encrypted_secret_bytes,
+        )
+
+    def test_tools_to_credential(self):
+        id_object_len = 40
+        encrypted_secret_len = 68
+
+        id_object_bytes = os.urandom(id_object_len)
+        encrypted_secret_bytes = os.urandom(encrypted_secret_len)
+
+        data = b""
+        data = data + int(0xBADCC0DE).to_bytes(4, "big") + int(1).to_bytes(4, "big")
+        data = data + id_object_len.to_bytes(2, "big") + id_object_bytes
+        data = data + encrypted_secret_len.to_bytes(2, "big") + encrypted_secret_bytes
```

Comment:
Use a static data structure for the tools crediantial structure. I wen't ahead and converted the tpm2-tools makecredential output to a python byte array you can shove into the test (see below). The reason for this is so we test that given a valid tpm2-tools output, you can parse that and we don't inject any randomness into the test using urandom and the test here only converts all within the python code, we're not using a known good tpm2-tools output.

```python3
mkcred = b"""
\xba\xdc\xc0\xde\x00\x00\x00\x01\x00\x2d\x00\x20\xbb\xc1\x84\x9f
\xec\x7a\xd5\x3f\x5f\x22\x33\xcb\xa3\x52\x03\x0a\xa7\x2e\x63\xc4
\xfc\x46\x50\x5f\x07\x10\xe3\x32\xc3\xb1\x71\x73\xe1\xb9\x22\x50
\xbd\x23\x7c\xe7\x03\x9d\x0b\x01\x00\x48\xa8\xaa\x1c\xf7\xfa\xdd
\xbb\xc0\xa0\xce\xf1\x2a\xb8\x5f\x4d\xac\xbc\xb8\xbb\x27\x71\x40
\xe2\x08\xb0\xba\x15\x11\xaf\x9d\x13\xeb\xc0\xf5\xa5\x66\x53\x62
\xad\x9a\x6c\xdc\x2e\x44\xe5\xd1\x9d\xa8\xb2\x67\x79\xec\x31\x05
\x8b\xc3\x93\x85\x81\x84\x21\xa9\xb3\xde\x28\x4e\xeb\x91\xba\xd5
\x35\x53\x8f\x10\x33\xc5\x2e\xe1\x05\x87\x6a\xa0\x17\xed\x21\xa6
\x66\x6b\x07\x54\x2f\x24\x26\xd6\xbd\x8b\x7b\xfa\x2f\x1b\x43\x4f
\x1e\x48\xec\x92\x44\x6e\xb6\xdc\x91\x2a\x92\x2f\x72\xbb\xf5\x2e
\xf2\x5b\xe9\xc0\xcb\x71\x96\xab\x13\x26\x78\xcb\xe2\x9e\x27\x16
\x17\x50\x5a\x7f\x6d\x73\x97\x8a\x68\xbf\xaf\x4d\x69\x10\x66\x74
\x60\x21\xa3\xc0\xf8\x15\xc0\x61\x65\x5b\x01\x16\x5b\x1a\x83\x72
\x9e\xbf\x41\x1a\x89\x83\x73\x89\xf9\xee\x40\xe5\xdf\x32\x49\xa1
\x3e\x04\x8d\xc0\x6e\x28\x4b\x00\xee\xeb\xa4\xcb\xbb\x65\x83\x87
\x26\x1f\xde\xd0\x17\x24\xa4\xc4\x99\x71\x48\xc0\x75\x89\x40\xce
\xaf\xcb\xf8\xfa\x64\xee\xf7\xb7\x4e\x32\x9d\x52\x7f\x1a\xc7\x95
\x8d\x40\xa0\xa5\xb6\x7d\x52\xfe\x94\x2f\xfd\xc3\x52\xef\x95\x64
\x42\x54\xd1\x80\xf4\xa2\xcd\xac\x66
"""
```

Here's the python code used to generate that block:
```python3
#!/usr/bin/env python3

import sys
with open("mkcred.out", "rb") as f:
    b = f.read()

print('mkcred = b"""')
for i, x in enumerate(b):
    if i > 0 and i % 16 == 0:
        sys.stdout.write("\n")
    sys.stdout.write(f"\\x{x:02x}")

sys.stdout.write('\n"""\n')
```

Reply:
It seems like the magic of the `mkcred` is invalid:
```
ValueError: bad magic, expected 0xBADCC0DE, got 0xABADCC0
```

Reply:
It seems like the magic of the `mkcred` is invalid:
```
ValueError: bad magic, expected 0xBADCC0DE, got 0xABADCC0
```

Reply:
The new lines in mkcred is included, that has caused me issues for other tests, check https://github.com/tpm2-software/tpm2-pytss/blob/97954eeed7d3a677b19929104391cffa7486c03c/test/test_types.py#L1504 for an ugly workaround

Reply:
Done!